Remove "impression"/"conversion" from API surfaces (#103)
WICG/conversion-measurement-api: README.md

-   **Advertiser**: Purchaser of ad slots, conversions happen on advertiser sites

> -   **Impression**: Click/view of an ad, equivalent to an attribution source.

Author: Removing sgtm

-   **Conversion**: The completion of a meaningful (advertiser specified) user action on the advertiser's web site by a user who has previously interacted with an ad from that advertiser.

-   **Event-level data**: Data that can be tied back to a specific low-level event; not aggregated

> -   **Click-through-conversion (CTC)**: A conversion attributed to an impression that was clicked

Author: Went with "ad click" as this is specific for click-through

Overview
========

Attribution Source Declaration
----------------------

### Registering attribution sources for anchor tag navigations

An attribution source is an anchor tag with special attributes:

> `<a attributeon="[eTLD+1]" attributionsourcedata="[string]"

Reviewer: [PCM issue](https://github.com/privacycg/private-click-measurement/issues/30#issuecomment-730598490): `ID`
[CM issue](https://github.com/WICG/conversion-measurement-api/issues/57#issuecomment-743403963): `data`

Which one has been chosen?

> `<a attributeon="[eTLD+1]" attributionsourcedata="[string]"

Author: Good question. I think originally we were planning on having separate attribute names for the click side data in PCM and CM API. PCM went with id so this changed to data.

However, per the decision on https://github.com/privacycg/private-click-measurement/issues/59, it seems like we are going to have different endpoints. Because the endpoints are separate, we could choose to use the same attribute name, attributionsourceid.

Note that attributionsourcedata matches up somewhat more nicely with the aggregate API equivalents, as they do not store IDs.

+csharrison do you have any thoughts here?

> `<a attributeon="[eTLD+1]" attributionsourcedata="[string]"

Reviewer 2: We should not use the same ID as PCM because otherwise it is difficult to interoperate (forces branching between the two browsers). Ideally we could get PCM to use "data" since it implies less entropy to me. Otherwise I liked "attribution-event-id" which was discussed in https://github.com/privacycg/private-click-measurement/issues/30

attributionexpiry=[unsigned long long] attributionreportto="[origin]">`

Attribution source attributes:

-   `attributeon`: the intended eTLD+1 destination of the ad click.

-   `attributionsourcedata`: the event-level data associated with this source. This will be limited to 64 bits of information, [encoded as a hexadecimal string](#data-encoding), but the value can vary for browsers that want a higher level of privacy.

-   `attributionexpiry`: (optional) expiry in milliseconds for when the source should be deleted. Default is 30 days, with a maximum value of 30 days. The maximum expiry can also vary between browsers.

-   `attributionreportto`: (optional) the desired endpoint that the attribution report for this source should go to. Default is the top level origin of the page.

Clicking on an anchor tag that specifies these attributes will create a new attribution source that will be handled according to [Handling an attribution source](#handling-an-attribution-source)

### Registering sources for window.open() navigations

> An attribution source can be registered for navigations initiated by [`window.open()`](https://html.spec.whatwg.org/multipage/window-object.html#dom-open).

Reviewer: @johnivdel EDIT: edited my comment above, made a typo on the option number

navigated to ends up sharing the an eTLD+1 with the `attributeon` attribute of
the attribution source. Concretely, this logs <`attributionsourcedata`,
`attributeon`, `attributionreportto`, `attributionexpiry`> to a new
browser storage area.

When an attribution source is logged for <`attributionreportto`,
`attributeon`>, existing sources matching this pair will be
looked up in storage. If the matching sources have been triggered at
least once (i.e. have scheduled a report), they will be removed from
browser storage and will not be eligible for further reporting. Any
pending reports for these sources will still be sent.

An attribution source will be eligible for reporting if any page on the	
`attributeon` domain (advertiser site) triggers attribution for the	
associated reporting origin.

### Publisher-side Controls for Source Declaration

In order to prevent arbitrary third parties from registering sources without the publisher’s knowledge, the Attribution Reporting API will need to be enabled in child contexts by a new [Feature Policy](https://w3c.github.io/webappsec-feature-policy/):

```
<iframe src="https://advertiser.test" allow="attribution-reporting ‘src’">

> <a … id="impressionTag" attributionreportto="https://ad-tech.com"></a>

Reviewer: (that was the last occurence of `impression` in this doc)

</iframe>
```

The API will be enabled by default in the top-level context and in same-origin children. Any script running in these contexts can declare a source with any reporting origin. Publishers who wish to explicitly disable the API for all parties can do so via an [HTTP header](https://w3c.github.io/webappsec-feature-policy/#feature-policy-http-header-field).

> Without a Feature Policy, a top-level document and cooperating iframe could recreate this functionality. This is possible by using [postMessage](https://html.spec.whatwg.org/multipage/web-messaging.html#dom-window-postmessage) to send the attribution source data, reportto origin, and attributeon origin to the top level document who can then wrap the iframe in an anchor tag (with some additional complexities behind handling clicks on the iframe). Using Feature Policy prevents the need for these hacks. This is inline with the classification of powerful features as discussed on [this issue](https://github.com/w3c/webappsec-feature-policy/issues/252).

Reviewer 2: Optional: Hm wonder if we should say "attributionreportto" rather than just "reportto"

Triggering Attribution
-----------------------

This API will use a mechanism for triggering attribution similar to the
[Ad Click Attribution Proposal](https://wicg.github.io/ad-click-attribution/index.html#legacytriggering).

Attribution triggering is meant to occur on `attributeon` pages. Attribution is

> triggered for a given `reportto` origin on a given `attributeon` through an HTTP GET to

Reviewer: Or `domain`

the origin, that redirects to a [.well-known](https://tools.ietf.org/html/rfc5785)
location.
This redirect is useful, because this mechanism enables the reporting origin to make server-side decisions about when attribution reports should trigger.
Note that `.well-known` is only used to register a path that the browser will understand; it shouldn't point to any actual resource, since the request will be cancelled internally.

Triggering attribution requires the `attribution-reporting` Feature Policy to be enabled in the context the request is made. As described in [Publisher Controls for Source Declaration](#publisher-side-controls-for-source-declaration), this Feature Policy will be enabled by default in the top-level context and in same-origin children, but disabled in cross-origin children.

Today, conversion pixels are frequently used to register conversions on
advertiser pages. These can be repurposed to trigger attribution in
this API:

```
<img src="https://ad-tech.test/conversiontracker"/>
```
`https://ad-tech.test/conversiontracker` can be redirected to `https://ad-tech.test/.well-known/trigger-attribution`

> to trigger a attribution for all .

Reviewer: `s/a/an`
+Is maybe a bit missing after `all`?

support fractions of bits, as it’s possible to limit data to values
from 0-5 (~2.6 bits of information)

> ### Trigger attribution algorithm

Reviewer: ```suggestion
### Attribution trigger algorithm
```

Reviewer: ?Not sure what sounds best, this is a nit

> When the browser receives a attriubtion trigger redirect on a URL matching

Reviewer: ```suggestion
When the browser receives a attribution trigger redirect on a URL matching
```

the `attributeon` eTLD+1, it looks up all sources in storage that
match <`attributionreportto`, `attributeon`>.

> The most recent matching source is given a `credit` of value 100. All other matching sources are given a `credit` of value of 0.

Reviewer: ```suggestion
The most recent matching source is given a `credit` of value 100. All other matching sources are given a `credit` of value 0.
```

For each matching source, schedule a report. To schedule a report,
the browser will store
 {`attributionreportto`, `attributeon` domain, `attributionsourcedata`, [decoded](#data-encoding) trigger-data, credit} for the source.
Scheduled reports will be sent as detailed in [Sending scheduled reports](#sending-scheduled-reports).

Each source is only allowed to schedule a maximum of three reports
(see [Triggering attribution multiple times for the same source](#triggering-attribution-multiple-times-for-the-same-source)). Once
reports are scheduled, the browser will delete all sources that have scheduled three reports.

### Multiple sources for the same trigger (Multi-touch)

If multiple sources were clicked and associated with a single attribution trigger, send reports for all of them. 

> To provide additional utility, the browser can choose to provide additional annotations to each of these reports, attributing credits for the triggering refirect to them individually. Attribution models allow for more sophisticated, accurate measurement.

Reviewer: ```suggestion
To provide additional utility, the browser can choose to provide additional annotations to each of these reports, attributing credits for the triggering redirect to them individually. Attribution models allow for more sophisticated, accurate measurement.
```

The default attribution model will be last-click attribution, giving the last-clicked source for a given trigger redirect all of the credit.

To remain flexible, the browser sends an `credit` of value 0 to 100 for all reports associated with a single trigger. This represents the percent of attribution a source received. The sum of credits across a set of reports for one trigger should equal 100.

There are many possible alternatives to this,
like providing a choice of rules-based attribution models. However, it
isn’t clear the benefits outweigh the additional complexity. Additionally, models other than last-click potentially leak more
cross-site information if sources are clicked across different
sites.

### Triggering attribution multiple times for the same source

Many ad clicks end up converting multiple times, for instance if a user
goes through a checkout and a purchase flow. To support this in a

> privacy preserving way, we need to make sure that subsequent triggering a source multiple times

Reviewer: ```suggestion
privacy preserving way, we need to make sure that triggering a source multiple times
```

> do not leak too much data.

Reviewer 2: ```suggestion
does not leak too much data.
```

One possible solution, outlined in this document, is for browsers to specify

> a maximum number of reports a single source can send. In this document

Reviewer: ```suggestion
a maximum number of reports that can send for a given source. In this document
```

Reviewer: To avoid confusion ("_the source sends the report_")

our initial proposal is 3.

Note that triggering attribution for the same source multiple times does not refresh
the reporting windows (see [Sending Scheduled Reports](#sending-scheduled-reports)).

Note that from a usability perspective, it is important that all
reports for the same source are allowed the same amount
of data. Otherwise, it becomes quite difficult for advertisers to
efficiently use the space of possible data values.

Sending Scheduled Reports
-------------------------

> After an attribution source is succesfully registered, a

Reviewer: ```suggestion
After an attribution source is successfully registered, a
```

schedule of reporting windows and deadlines associated with that
source begins. The time between the click and expiry can
be split into multiple reporting windows, at the end of which the
browser will send scheduled reports for that source.

Each reporting window has a deadline, and only reports created
before that deadline can be sent in that window. An example of deadlines
and windows a browser could choose are:

> 2 days minus 1 hour: Reports will be sent 2 days from source registration

Reviewer: Nice, this reads much simpler ✨

Note that to improve utility, it might be possible to randomly send
reports throughout each reporting window.

### Attribution Reports

To send a report, the browser will make a non-credentialed (i.e. without session cookies) secure
HTTP POST request to:

```
https://attributionreportto/.well-known/report-attribution
```

The report data is included in the request body as a JSON object with the following keys:

> -   `sourceData`: 64 bit data set on the attribution source

Reviewer: Also just to check: the JSON keys will be camelCased?

The report data is included in the request body as a JSON object with the following keys:

> -   `source_data`: 64 bit data set on the attribution source

Reviewer 2: Should be source_event_id

Author: Done

`toasters.com`, and sets `ad-tech.com` to be an allowed reporting origin.

Within the iframe, `toasters.com` code annotates their anchor tags to use
the `ad-tech.com` reporting origin, and uses a source data value that allows
`ad-tech.com` to identify the ad click (12345678)
```
<iframe src="https://ad-tech-3p.test/show-some-ad" allow="attribution-reporting ‘src’ (https://ad-tech.com)">
...
<a 
  href="https://toasters.com/purchase"
  attributeon="https://toasters.com"
  attributionsourcedata="12345678"
  attributionreportto="https://ad-tech.com"
  attributionexpiry=604800000>
...
</iframe>
```

A user clicks on the ad and this opens a window that lands on a URL to
`toasters.com/purchase`. An attribution source is logged to browser storage

> since the landing page matches the `attributeon` etld+1. The following data is

Reviewer: ```suggestion
since the landing page matches the `attributeon` eTLD+1. The following data is
```

ads on, including `ad-tech.com`, by adding conversion pixels:

```
<img src="https://ad-tech.com/trigger-attribution?model=toastmaster3000&price=$49.99&..." />
```

`ad-tech.com` receives this request, and decides to trigger attribution
on `toasters.com`. They must compress all of the data into
3 bits, so `ad-tech.com` chooses to encode the value as “2" (e.g. some
bucketed version of the purchase value). They respond with a 302
redirect to:
```
https://ad-tech.com/.well-known/trigger-attribution?data=2
```

The browser sees this request, and schedules a report to be
sent. The report is associated with the 7 day deadline as the
2 day deadline has passed. Roughly 5 days later, `ad-tech.com` receives
the following HTTP POST:
```
URL:
https://ad-tech.com/.well-known/report-attribution

body:
{
  "sourceData": "12345678",
  "triggerData": "2",
  "creidt": 100
}
```

Privacy Considerations
======================
The main privacy goal of the API is to make _linking identity_ between two different top-level sites difficult. This happens when either a request or a Javascript environment has two user IDs from two different sites simultaneously.

> In this API, the 64-bit source ID can encode a user ID from the publisher’s top level site, but the low entropy, noisy trigger data could only encode a small part of a user ID from the advertiser’s top-level site. The source ID and the trigger data are never exposed to a Javascript environment together, and the request that includes both of them is sent without credentials and at a different time from either event, so the request adds little new information linkable to these events.

Reviewer: [PCM issue](https://github.com/privacycg/private-click-measurement/issues/30#issuecomment-730598490): `ID`
[CM issue](https://github.com/WICG/conversion-measurement-api/issues/57#issuecomment-743403963): `data`

Which one has been chosen?

While this API _does_ allow you to learn "which ad clicks converted", it isn’t enough to link the user's identity on the publisher's and advertiser's side, unless there is serious abuse of the API, i.e. abusers are using error correcting codes and many clicks to slowly and probabilistically learn advertiser IDs associated with publisher ones. We explore some mitigations to this attack below.

Trigger Data
-------------------

> Trigger data, e.g. advertiser side data, is extremely important for critical use cases like

Reviewer: ```suggestion
Trigger data, e.g. advertiser-side data, is extremely important for critical use cases like
```

reporting the *purchase value* of a conversion. However, too much advertiser-side
data could be used to link advertiser identity with publisher
identity.

Mitigations against this are to provide only coarse information (only a

> few bits at a time), and introduce some noise to the trigger-data. Even

Reviewer: ```suggestion
few bits at a time), and introduce some noise to the trigger data. Even
```

{

>   "sourceData": "12345678",

Reviewer 2: Update to use underscores, source_event_data

Author: Done

[Randomized response](https://en.wikipedia.org/wiki/Randomized_response) for
a starting point.

Reporting Delay 
-----------------

By bucketing reports within a small number reporting deadlines, it
becomes harder to associate a report with the identity of the
user on the advertiser’s site via timing side channels.

> Reprots within the same reporting window occur within an anonymity

Reviewer: ```suggestion
Reports within the same reporting window occur within an anonymity
```
